src/pool/srv_pool.c

```diff
@@ -5081,64 +5188,116 @@ pool_svc_reconfigure_replicas(struct pool_svc *svc, int svc_rf, struct pool_map
 	 */
 	if (to_add->rl_nr > 0)
 		ds_rsvc_add_replicas_s(&svc->ps_rsvc, to_add, ds_rsvc_get_md_cap());
+	if (reconf->psc_canceled)
+		goto out_to_add_remove;
 	if (to_add->rl_nr > to_remove->rl_nr)
 		to_remove->rl_nr = 0;
 	else
 		to_remove->rl_nr -= to_add->rl_nr;
-	if (to_remove->rl_nr > 0)
-		ds_rsvc_remove_replicas_s(&svc->ps_rsvc, to_remove, false /* stop */);
+	if (to_remove->rl_nr > 0) {
+		d_rank_list_t *tmp;
+
+		/*
+		 * Since the ds_rsvc_dist_stop part is likely to hit RPC
+		 * timeouts, after removing the replicas from the membership,
+		 * we notify the MS first, and then come back to
+		 * ds_rsvc_dist_stop.
+		 */
+		rc = d_rank_list_dup(&tmp, to_remove);
+		if (rc != 0) {
+			D_ERROR(DF_UUID": failed to duplicate to_remove: "DF_RC"\n",
+				DP_UUID(svc->ps_uuid), DP_RC(rc));
+			goto out_to_add_remove;
+		}
+		rc = rdb_remove_replicas(svc->ps_rsvc.s_db, tmp);
+		/* Delete from to_remove ranks that are not removed. */
+		d_rank_list_filter(tmp, to_remove, true /* exclude */);
+		d_rank_list_free(tmp);
+	}
 
 	if (rdb_get_ranks(svc->ps_rsvc.s_db, &new) == 0) {
 		d_rank_list_sort(current);
 		d_rank_list_sort(new);
 
-		if (!d_rank_list_identical(new, current)) {
+		if (reconf->psc_force_notify || !d_rank_list_identical(new, current)) {
```

Comment:
I may be getting a little lost looking at all of the parts of this change, but a question: we don't expect many instances where the psc_force_notify will cause a RAS notification even when the lists are identical (no membership changes occur), is that right? Since most of the time if we are here, there is some membership change occurring (e.g., you would be stepping up as a leader because something happened to the previous leader)?

Reply:
I was thinking about a scenario where a new leader steps up but the membership has not changed - so I guess now looking closer maybe this could be the consequence of the previous leader experiencing some failure (e.g., in a step of updating the pool map) and calling rdb_resign() due to that error.

But I see your line of thinking here, that if for some very unexpected reason ds_notify_pool_svc_update() fails then reconf->psc_force_notify will remain true.

Probably no issue here, as we will want to see any leader transition reported.